File: src/shared/runtime.py
from __future__ import annotations

import json
import logging
import math
import re
import time
from collections.abc import Iterable, Mapping
from datetime import UTC, datetime
from typing import Any, Literal, cast

from pydantic import Field, field_validator
from pydantic_settings import BaseSettings, SettingsConfigDict

_LOG_FIELDS = (
    "request_id",
    "provider",
    "tool",
    "endpoint",
    "duration_ms",
    "status_code",
    "retry_count",
    "cache_hit",
)
_MAX_LOG_MESSAGE_LENGTH = 2048
_SENSITIVE_KEY = re.compile(
    r"(?i)(?:private[-_ ]?key|password|passwd|token|auth|api[-_ ]?key|secret|cookie)"
)
_SENSITIVE_MESSAGE = re.compile(
    r'(?i)(?P<prefix>\b(?:private[-_ ]?key|authorization|auth|api[-_ ]?key|password|'
    r'passwd|secret|token|access[-_ ]?token|refresh[-_ ]?token|'
    r'client[-_ ]?secret|cookie)\b\s*["\']?\s*[:=]\s*)'
    r'(?:(?P<double>"(?:\\.|[^"\\])*")|(?P<single>\'(?:\\.|[^\'\\])*\')|'
    r'(?P<bare>(?:bearer\s+)?[^\s,;}\]]+))'
)

logger = logging.getLogger(__name__)


def _redact_json_value(value: Any) -> Any:
    if isinstance(value, dict):
        mapping = cast(dict[str, Any], value)
        redacted: dict[str, Any] = {}
        for key, item in mapping.items():
            redacted[key] = "[REDACTED]" if _SENSITIVE_KEY.search(key) else _redact_json_value(item)
        return redacted
    if isinstance(value, list):
        items = cast(list[Any], value)
        return [_redact_json_value(item) for item in items]
    if isinstance(value, str):
        nested = _redact_json_message(value)
        return nested if nested is not None else value
    return value


def _redact_json_message(message: str) -> str | None:
    decoder = json.JSONDecoder()
    cursor = 0
    parts: list[str] = []
    found = False
    while cursor < len(message):
        object_start = message.find("{", cursor)
        array_start = message.find("[", cursor)
        starts = [start for start in (object_start, array_start) if start >= 0]
        if not starts:
            break
        start = min(starts)
        try:
            value, end = decoder.raw_decode(message, start)
        except json.JSONDecodeError:
            cursor = start + 1
            continue
        if not isinstance(value, (dict, list)):
            cursor = end
            continue
        redacted = json.dumps(
            _redact_json_value(value), ensure_ascii=True, separators=(",", ":")
        )
        parts.extend((message[cursor:start], redacted))
        cursor = end
        found = True
    if not found:
        return None
    parts.append(message[cursor:])
    return "".join(parts)


def _safe_message(message: str) -> str:
    def redact(match: re.Match[str]) -> str:
        value = match.group("double") or match.group("single")
        quote = value[0] if value is not None else ""
        return f'{match.group("prefix")}{quote}[REDACTED]{quote}'

    redacted = _redact_json_message(message) or message
    redacted = _SENSITIVE_MESSAGE.sub(redact, redacted)
    if len(redacted) <= _MAX_LOG_MESSAGE_LENGTH:
        return redacted
    return f"{redacted[: _MAX_LOG_MESSAGE_LENGTH - 3]}..."


def _sanitize_extra(value: Any, key: str | None = None) -> Any:
    if key is not None and _SENSITIVE_KEY.search(key):
        return "[REDACTED]"
    if isinstance(value, str):
        return _safe_message(value)
    if isinstance(value, Mapping):
        mapping = cast(Mapping[Any, Any], value)
        sanitized: dict[str, Any] = {}
        for raw_key, item in mapping.items():
            item_key = str(raw_key)
            sanitized[item_key] = _sanitize_extra(item, item_key)
        return sanitized
    if isinstance(value, (list, tuple, set, frozenset)):
        items = cast(Iterable[Any], value)
        return [_sanitize_extra(item) for item in items]
    if value is None or isinstance(value, (bool, int, float)):
        return value
    return _safe_message(str(value))


class _JsonFormatter(logging.Formatter):
    def format(self, record: logging.LogRecord) -> str:
        payload: dict[str, object] = {
            "timestamp": datetime.now(UTC).isoformat(timespec="milliseconds").replace(
                "+00:00", "Z"
            ),
            "level": _sanitize_extra(record.levelname),
            "name": _sanitize_extra(record.name),
            "message": _safe_message(record.getMessage()),
        }
        for field in _LOG_FIELDS:
            if field in record.__dict__:
                payload[field] = _sanitize_extra(record.__dict__[field], field)
        return json.dumps(payload, ensure_ascii=True, default=str)


def configure_logging(settings: Settings | None = None) -> None:
    configuration = settings if settings is not None else Settings()  # pyright: ignore[reportCallIssue]
    level = logging.getLevelNamesMapping().get(configuration.log_level.upper(), logging.INFO)
    logger.setLevel(level)
    logger.propagate = False

    handler = next(
        (
            candidate
            for candidate in logger.handlers
            if getattr(candidate, "_compras_json_handler", False)
        ),
        None,
    )
    if handler is None:
        handler = logging.StreamHandler()
        handler._compras_json_handler = True  # type: ignore[attr-defined]
        logger.addHandler(handler)
    handler.setLevel(logging.NOTSET)
    handler.setFormatter(_JsonFormatter())


class Settings(BaseSettings):
    model_config = SettingsConfigDict(env_file=".env", extra="ignore")

    compras_base_url: str = "https://dadosabertos.compras.gov.br"
    pncp_base_url: str = "https://pncp.gov.br/api/pncp"
    http_timeout: float = Field(30, gt=0)
    http_max_retries: int = Field(3, gt=0)
    http_requests_per_second: int = Field(5, gt=0)
    compras_max_concurrency: int = Field(4, gt=0)
    pncp_max_concurrency: int = Field(4, gt=0)
    cache_enabled: bool = True
    cache_domains_ttl: int = Field(86400, gt=0)
    cache_catalog_ttl: int = Field(3600, gt=0)
    cache_recent_ttl: int = Field(300, gt=0)
    cache_historical_ttl: int = Field(86400, gt=0)
    max_document_bytes: int = Field(25_000_000, gt=0)
    mcp_transport: Literal["stdio", "http"] = "stdio"
    log_level: str = "INFO"

    @field_validator(
        "http_max_retries",
        "http_requests_per_second",
        "compras_max_concurrency",
        "pncp_max_concurrency",
        "cache_domains_ttl",
        "cache_catalog_ttl",
        "cache_recent_ttl",
        "cache_historical_ttl",
        "max_document_bytes",
        mode="before",
    )
    @classmethod
    def _reject_non_integer_settings(cls, value: Any) -> Any:
        if isinstance(value, (bool, float)):
            raise ValueError("must be an integer")
        return value

    @field_validator("http_timeout", mode="before")
    @classmethod
    def _reject_boolean_timeout(cls, value: Any) -> Any:
        if isinstance(value, bool):
            raise ValueError("must be numeric seconds")
        return value

    @field_validator("http_timeout")
    @classmethod
    def _reject_non_finite_timeout(cls, value: float) -> float:
        if not math.isfinite(value):
            raise ValueError("must be finite")
        return value


class TtlCache:
    def __init__(self, max_entries: int = 4096) -> None:
        if max_entries <= 0:
            raise ValueError("max_entries must be positive")
        self._max_entries = max_entries
        self._entries: dict[str, tuple[float, Any]] = {}

    def get(self, key: str) -> Any | None:
        entry = self._entries.get(key)
        if entry is None or entry[0] <= time.monotonic():
            self._entries.pop(key, None)
            return None
        return entry[1]

    def put(self, key: str, value: Any, ttl_seconds: int) -> None:
        if ttl_seconds <= 0:
            raise ValueError("ttl_seconds must be positive")
        if key not in self._entries and len(self._entries) >= self._max_entries:
            self._entries.pop(next(iter(self._entries)))
        self._entries[key] = (time.monotonic() + ttl_seconds, value)

File: src/shared/runtime.ts
import { performance } from "node:perf_hooks";

const LOG_FIELDS: readonly string[] = [
  "request_id",
  "provider",
  "tool",
  "endpoint",
  "duration_ms",
  "status_code",
  "retry_count",
  "cache_hit",
];
const MAX_LOG_MESSAGE_LENGTH = 2048;
const SENSITIVE_KEY =
  /(?:private[-_ ]?key|password|passwd|token|auth|api[-_ ]?key|secret|cookie)/i;
const SENSITIVE_MESSAGE =
  /(?<prefix>\b(?:private[-_ ]?key|authorization|auth|api[-_ ]?key|password|passwd|secret|token|access[-_ ]?token|refresh[-_ ]?token|client[-_ ]?secret|cookie)\b\s*["']?\s*[:=]\s*)(?:(?<double>"(?:\\.|[^"\\])*")|(?<single>'(?:\\.|[^'\\])*')|(?<bare>(?:bearer\s+)?[^\s,;}\]]+))/gi;

export interface Settings {
  comprasBaseUrl: string;
  pncpBaseUrl: string;
  httpTimeout: number;
  httpMaxRetries: number;
  httpRequestsPerSecond: number;
  comprasMaxConcurrency: number;
  pncpMaxConcurrency: number;
  cacheEnabled: boolean;
  cacheDomainsTtl: number;
  cacheCatalogTtl: number;
  cacheRecentTtl: number;
  cacheHistoricalTtl: number;
  maxDocumentBytes: number;
  mcpTransport: "stdio" | "http";
  logLevel: string;
}

function redactJsonValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(redactJsonValue);
  }
  if (typeof value === "object" && value !== null) {
    const redacted: Record<string, unknown> = {};
    for (const [key, item] of Object.entries(value)) {
      redacted[key] = SENSITIVE_KEY.test(key) ? "[REDACTED]" : redactJsonValue(item);
    }
    return redacted;
  }
  if (typeof value === "string") {
    return redactJsonMessage(value) ?? value;
  }
  return value;
}

// ponytail: bracket-matching scanner emulating json raw_decode; upgrade path is
// a real incremental JSON parser only if pathological inputs show up.
function parseJsonAt(text: string, start: number): [unknown, number] | null {
  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let i = start; i < text.length; i++) {
    const ch = text[i]!;
    if (inString) {
      if (escaped) escaped = false;
      else if (ch === "\\") escaped = true;
      else if (ch === '"') inString = false;
      continue;
    }
    if (ch === '"') inString = true;
    else if (ch === "{" || ch === "[") depth++;
    else if (ch === "}" || ch === "]") {
      depth--;
      if (depth === 0) {
        try {
          return [JSON.parse(text.slice(start, i + 1)), i + 1];
        } catch {
          return null;
        }
      }
    } else if (depth === 0) {
      return null;
    }
  }
  return null;
}

function redactJsonMessage(message: string): string | null {
  let cursor = 0;
  const parts: string[] = [];
  let found = false;
  while (cursor < message.length) {
    const objectStart = message.indexOf("{", cursor);
    const arrayStart = message.indexOf("[", cursor);
    const starts = [objectStart, arrayStart].filter((start) => start >= 0);
    if (starts.length === 0) break;
    const start = Math.min(...starts);
    const parsed = parseJsonAt(message, start);
    if (parsed === null) {
      cursor = start + 1;
      continue;
    }
    const [value, end] = parsed;
    if (typeof value !== "object" || value === null) {
      cursor = end;
      continue;
    }
    parts.push(message.slice(cursor, start), JSON.stringify(redactJsonValue(value)));
    cursor = end;
    found = true;
  }
  if (!found) return null;
  parts.push(message.slice(cursor));
  return parts.join("");
}

function safeMessage(message: string): string {
  let redacted = redactJsonMessage(message) ?? message;
  redacted = redacted.replace(
    SENSITIVE_MESSAGE,
    (...args: unknown[]) => {
      const groups = args[args.length - 1] as {
        prefix: string;
        double?: string;
        single?: string;
      };
      const value = groups.double || groups.single;
      const quote = value !== undefined ? value[0] : "";
      return `${groups.prefix}${quote}[REDACTED]${quote}`;
    },
  );
  if (redacted.length <= MAX_LOG_MESSAGE_LENGTH) return redacted;
  return `${redacted.slice(0, MAX_LOG_MESSAGE_LENGTH - 3)}...`;
}

function sanitizeExtra(value: unknown, key?: string): unknown {
  if (key !== undefined && SENSITIVE_KEY.test(key)) return "[REDACTED]";
  if (typeof value === "string") return safeMessage(value);
  if (Array.isArray(value)) return value.map((item) => sanitizeExtra(item));
  if (value === null || typeof value === "boolean" || typeof value === "number") {
    return value;
  }
  if (typeof value === "object") {
    const sanitized: Record<string, unknown> = {};
    for (const [rawKey, item] of Object.entries(value)) {
      sanitized[rawKey] = sanitizeExtra(item, rawKey);
    }
    return sanitized;
  }
  return safeMessage(String(value));
}

const LEVELS: Record<string, number> = {
  debug: 10,
  info: 20,
  warning: 30,
  error: 40,
  critical: 50,
};

let minLevel = LEVELS["info"]!;

export function configureLogging(settings?: Settings): void {
  const name = (settings?.logLevel ?? "INFO").toLowerCase();
  minLevel = LEVELS[name] ?? LEVELS["info"]!;
}

function log(level: "INFO" | "WARNING" | "ERROR", numeric: number, message: string, extra?: Record<string, unknown>): void {
  if (numeric < minLevel) return;
  const payload: Record<string, unknown> = {
    timestamp: new Date().toISOString(),
    level,
    name: "compras-publicas-br",
    message: safeMessage(message),
  };
  if (extra !== undefined) {
    for (const field of LOG_FIELDS) {
      if (Object.hasOwn(extra, field)) {
        payload[field] = sanitizeExtra(extra[field], field);
      }
    }
  }
  process.stderr.write(`${JSON.stringify(payload)}\n`);
}

export const logger = {
  info(message: string, extra?: Record<string, unknown>): void {
    log("INFO", 20, message, extra);
  },
  warn(message: string, extra?: Record<string, unknown>): void {
    log("WARNING", 30, message, extra);
  },
  error(message: string, extra?: Record<string, unknown>): void {
    log("ERROR", 40, message, extra);
  },
};

let envFileLoaded = false;

function parsePositiveInt(env: NodeJS.ProcessEnv, name: string, fallback: number): number {
  const raw = env[name];
  if (raw === undefined) return fallback;
  const value = Number(raw);
  if (!Number.isFinite(value) || !Number.isInteger(value)) {
    throw new Error(`${name}: must be an integer`);
  }
  if (value <= 0) {
    throw new Error(`${name}: Input should be greater than 0`);
  }
  return value;
}

export function loadSettings(env: NodeJS.ProcessEnv = process.env): Settings {
  if (!envFileLoaded) {
    envFileLoaded = true;
    try {
      const loadEnvFile = (
        process as unknown as { loadEnvFile?: (path: string) => void }
      ).loadEnvFile;
      loadEnvFile?.call(process, ".env");
    } catch {
      // ignore missing/invalid .env
    }
  }

  const timeoutRaw = env["HTTP_TIMEOUT"];
  let httpTimeout = 30;
  if (timeoutRaw !== undefined) {
    if (timeoutRaw === "true" || timeoutRaw === "false") {
      throw new Error("HTTP_TIMEOUT: must be numeric seconds");
    }
    httpTimeout = Number(timeoutRaw);
    if (Number.isNaN(httpTimeout)) {
      throw new Error("HTTP_TIMEOUT: must be numeric seconds");
    }
    if (!Number.isFinite(httpTimeout)) {
      throw new Error("HTTP_TIMEOUT: must be finite");
    }
    if (httpTimeout <= 0) {
      throw new Error("HTTP_TIMEOUT: Input should be greater than 0");
    }
  }

  const cacheEnabledRaw = (env["CACHE_ENABLED"] ?? "true").toLowerCase();
  let cacheEnabled: boolean;
  if (cacheEnabledRaw === "true") cacheEnabled = true;
  else if (cacheEnabledRaw === "false") cacheEnabled = false;
  else throw new Error("CACHE_ENABLED: Input should be a valid boolean");

  const mcpTransportRaw = env["MCP_TRANSPORT"] ?? "stdio";
  if (mcpTransportRaw !== "stdio" && mcpTransportRaw !== "http") {
    throw new Error("MCP_TRANSPORT: Input should be 'stdio' or 'http'");
  }

  return {
    comprasBaseUrl: env["COMPRAS_BASE_URL"] ?? "https://dadosabertos.compras.gov.br",
    pncpBaseUrl: env["PNCP_BASE_URL"] ?? "https://pncp.gov.br/api/pncp",
    httpTimeout,
    httpMaxRetries: parsePositiveInt(env, "HTTP_MAX_RETRIES", 3),
    httpRequestsPerSecond: parsePositiveInt(env, "HTTP_REQUESTS_PER_SECOND", 5),
    comprasMaxConcurrency: parsePositiveInt(env, "COMPRAS_MAX_CONCURRENCY", 4),
    pncpMaxConcurrency: parsePositiveInt(env, "PNCP_MAX_CONCURRENCY", 4),
    cacheEnabled,
    cacheDomainsTtl: parsePositiveInt(env, "CACHE_DOMAINS_TTL", 86400),
    cacheCatalogTtl: parsePositiveInt(env, "CACHE_CATALOG_TTL", 3600),
    cacheRecentTtl: parsePositiveInt(env, "CACHE_RECENT_TTL", 300),
    cacheHistoricalTtl: parsePositiveInt(env, "CACHE_HISTORICAL_TTL", 86400),
    maxDocumentBytes: parsePositiveInt(env, "MAX_DOCUMENT_BYTES", 25_000_000),
    mcpTransport: mcpTransportRaw,
    logLevel: env["LOG_LEVEL"] ?? "INFO",
  };
}

export class TtlCache {
  private readonly maxEntries: number;
  private readonly entries = new Map<string, { expiresAt: number; value: unknown }>();

  constructor(maxEntries = 4096) {
    if (maxEntries <= 0) {
      throw new Error("max_entries must be positive");
    }
    this.maxEntries = maxEntries;
  }

  get<T = unknown>(key: string): T | undefined {
    const entry = this.entries.get(key);
    if (entry === undefined || entry.expiresAt <= performance.now()) {
      this.entries.delete(key);
      return undefined;
    }
    return entry.value as T;
  }

  put(key: string, value: unknown, ttlSeconds: number): void {
    if (ttlSeconds <= 0) {
      throw new Error("ttl_seconds must be positive");
    }
    if (!this.entries.has(key) && this.entries.size >= this.maxEntries) {
      const oldest = this.entries.keys().next().value as string;
      this.entries.delete(oldest);
    }
    this.entries.set(key, { expiresAt: performance.now() + ttlSeconds * 1000, value });
  }
}
